crates/wasi/src/filesystem.rs

```diff
@@ -170,30 +186,47 @@ impl Dir {
         F: FnOnce(&cap_std::fs::Dir) -> R + Send + 'static,
         R: Send + 'static,
     {
-        let d = self.dir.clone();
-        spawn_blocking(move || body(&d)).await
+        if self.allow_blocking_current_thread {
+            body(&self.dir)
+        } else {
+            let d = self.dir.clone();
+            spawn_blocking(move || body(&d)).await
+        }
     }
 }
 
 pub struct FileInputStream {
     file: Arc<cap_std::fs::File>,
     position: u64,
+    allow_blocking_current_thread: bool,
 }
 impl FileInputStream {
-    pub fn new(file: Arc<cap_std::fs::File>, position: u64) -> Self {
-        Self { file, position }
+    pub fn new(
+        file: Arc<cap_std::fs::File>,
+        position: u64,
+        allow_blocking_current_thread: bool,
+    ) -> Self {
+        Self {
+            file,
+            position,
+            allow_blocking_current_thread,
+        }
     }
 
     pub async fn read(&mut self, size: usize) -> Result<Bytes, StreamError> {
         use system_interface::fs::FileIoExt;
-        let f = Arc::clone(&self.file);
         let p = self.position;
-        let (r, mut buf) = spawn_blocking(move || {
+        let run = move |f: &cap_std::fs::File| {
             let mut buf = BytesMut::zeroed(size);
             let r = f.read_at(&mut buf, p);
             (r, buf)
-        })
-        .await;
+        };
+        let (r, mut buf) = if self.allow_blocking_current_thread {
+            run(&self.file)
+        } else {
+            let f = Arc::clone(&self.file);
+            spawn_blocking(move || run(&f)).await
+        };
```

Comment:
Might be kinda nice to wrap this dance into a helper method so that we don't have to check `self.allow_blocking_current_thread` everywhere and instead call `self.spawn_blocking` or something that internally either does a real `spawn_blocking` or runs the thing immediately on the current thread.

Reply:
Sounds good yeah, I did some refactoring to thread this through `File` which has a similar helper already.